File: homeworks/hw20/t_20_13_e4082.py
from math import log2, gcd

class SegmentTree:
    def __init__(self, array):
        n = (1 << int(log2(len(array)-1))+1)
        self.size = n
        self.items = 2*n*[1]

        for i in range(len(array)):
            self.items[n + i] = array[i]

        for i in range(n-1, 0, -1):
            self.items[i] = self.items[2*i] * self.items[2*i+1]
    
    def update(self, pos, x):
        pos += self.size
        self.items[pos] = x
        i = pos // 2
        while i > 0:
            self.items[i] = self.items[2*i] * self.items[2*i+1]
            i //= 2
    
    def prod(self, left, right):
        left += self.size
        right += self.size
        res = 1
        while left <= right:
            if left % 2 == 1:
                res *= self.items[left]
            if right % 2 == 0:
                res *= self.items[right]
            
            left = (left + 1) // 2
            right = (right - 1) // 2
        return res

if __name__ == "__main__":
    with open("input.txt") as f:
        line = f.readline()
        while line:
            n, k = map(int, line.split())
            array = [int(el) for el in f.readline().split()]
            st = SegmentTree(array)
            for _ in range(k):
                cmd, *args = f.readline().split()
                if cmd == "P":
                    res = st.prod(int(args[0])-1, int(args[1])-1)
                    if res == 0:
                        print("0", end = "")
                    elif res > 0:
                        print("+", end = "")
                    else:
                        print("-", end = "")
                elif cmd == "C":
                    st.update(int(args[0])-1, int(args[1]))
            print()
            line = f.readline()
 

File: homeworks/hw20/t_20_13_e4082.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <cmath>

using namespace std;

int sign(int x) {
    if (x > 0) return 1;
    if (x < 0) return -1;
    return 0;
}

class SegmentTree {
public:
    int size;
    vector<int> items;

    SegmentTree(const vector<int>& array) {
        int n = 1 << ((int)log2(array.size() - 1) + 1);
        size = n;
        items.assign(2 * n, 1);

        for (int i = 0; i < array.size(); ++i) {
            items[n + i] = sign(array[i]);
        }

        for (int i = n - 1; i > 0; --i) {
            items[i] = items[2 * i] * items[2 * i + 1];
        }
    }

    void update(int pos, int x) {
        pos += size;
        items[pos] = sign(x);
        for (int i = pos / 2; i > 0; i /= 2) {
            items[i] = items[2 * i] * items[2 * i + 1];
        }
    }

    int prod(int left, int right) {
        left += size;
        right += size;
        int res = 1;
        while (left <= right) {
            if (left % 2 == 1) res *= items[left++];
            if (right % 2 == 0) res *= items[right--];
            if (res == 0) return 0;
            left /= 2;
            right /= 2;
        }
        return res;
    }
};

int main() {
    ifstream fin("input.txt");
    string line;

    while (getline(fin, line)) {
        if (line.empty()) continue;

        int n, k;
        sscanf(line.c_str(), "%d %d", &n, &k);

        vector<int> array(n);
        for (int i = 0; i < n; ++i) {
            fin >> array[i];
        }

        SegmentTree st(array);
        for (int i = 0; i < k; ++i) {
            char cmd;
            int a, b;
            fin >> cmd >> a >> b;
            if (cmd == 'P') {
                int res = st.prod(a - 1, b - 1);
                if (res == 0) cout << "0";
                else if (res > 0) cout << "+";
                else cout << "-";
            } else if (cmd == 'C') {
                st.update(a - 1, b);
            }
        }
        cout << endl;
        getline(fin, line);
    }

    return 0;
}
